src/Equinox.MessageDb/MessageDbClient.fs

```diff
@@ -13,72 +14,88 @@ open NpgsqlTypes
-                   (metadata::jsonb->>'$correlationId')::text,
-                   (metadata::jsonb->>'$causationId')::text,
-                   time
-                 from get_stream_messages(@StreamName, @FromPosition, @BatchSize)"
-
-            cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, streamName)
-            |> ignore
-
-            cmd.Parameters.AddWithValue("FromPosition", NpgsqlDbType.Bigint, fromPosition)
-            |> ignore
-
-            cmd.Parameters.AddWithValue("BatchSize", NpgsqlDbType.Bigint, batchSize)
-            |> ignore
-
-            use reader = cmd.ExecuteReader()
-
-            let! hasNext = reader.ReadAsync(ct)
-            let events = ResizeArray()
-            let mutable hasNext = hasNext
-
-            while hasNext do
-                let meta =
-                    if reader.IsDBNull(3) then
-                        "null"
-                    else
-                        reader.GetString(3)
-
-                let time =
-                    DateTime.SpecifyKind(reader.GetDateTime(7), DateTimeKind.Utc)
-
-                let timestamp = DateTimeOffset(time)
-
-                let timelineEvent =
-                    TimelineEvent.Create(
-                        index = reader.GetInt64(0),
-                        eventType = reader.GetString(1),
-                        data = JsonSerializer.Deserialize<JsonElement>(reader.GetString(2)),
-                        meta = JsonSerializer.Deserialize<JsonElement>(meta),
-                        eventId = reader.GetGuid(4),
-                        ?correlationId =
-                            (if reader.IsDBNull(5) then
-                                 None
-                             else
-                                 Some(reader.GetString(5))),
-                        ?causationId =
-                            (if reader.IsDBNull(6) then
-                                 None
-                             else
-                                 Some(reader.GetString(6))),
-                        timestamp = timestamp
-                    )
-
-                events.Add(timelineEvent)
-                let! next = reader.ReadAsync(ct)
-                hasNext <- next
-
-            return events.ToArray()
-        }
+    let readRow (reader: DbDataReader) =
+        let meta =
+            if reader.IsDBNull(3) then
+                "null"
+            else
+                reader.GetString(3)
+
+        let time =
+            DateTime.SpecifyKind(reader.GetDateTime(7), DateTimeKind.Utc)
+
+        let timestamp = DateTimeOffset(time)
+
+        TimelineEvent.Create(
+            index = reader.GetInt64(0),
+            eventType = reader.GetString(1),
+            data = JsonSerializer.Deserialize<JsonElement>(reader.GetString(2)),
+            meta = JsonSerializer.Deserialize<JsonElement>(meta),
+            eventId = reader.GetGuid(4),
+            ?correlationId =
+                (if reader.IsDBNull(5) then
```

Comment:
pull out a helper 
```
let tryGetString ord =
    if reader.IsDBNull ord then None
    else reader.GetString ord |> Some
 ```

Reply:
Done, I've also moved away from the JsonElement codec usage. Did not really have a good reason for it.

Reply:
Yeah, I guess if you were going to splice into it via JsonDocument stuff there are paths where it makes sense but it can take time to make these realizations (def good to use ROM as a default as that's what most stuff uses)
Soon we'll probably bring buffer pooling back from the FsCodec.NewtonsoftJson too